core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
          throw new PersistenceException("Could not add new merged entry", e);
       }
    }
+
    @Override
-   public void purge(Executor threadPool, final PurgeListener task) {
+   public Publisher<MarshallableEntry<K, V>> purgeExpired() {
       long now = timeService.wallClockTime();
-      List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
-      synchronized (entries) {
-         for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
-            Map.Entry<K, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
-            }
-         }
-      }
+      UnicastProcessor<MarshallableEntry<K, V>> processor = UnicastProcessor.create();
+      blockingManager.runBlocking(() -> {
+         blockingPurgeExpired(now, processor);
+      }, "sfs-purgeExpired");
+      return Flowable.fromPublisher(processor.serialize());
+   }
 
+   private void blockingPurgeExpired(long now, UnicastProcessor<MarshallableEntry<K, V>> processor) {
       long stamp = resizeLock.readLock();
       try {
-         for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
-            KeyValuePair<Object, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               try {
-                  free(fe);
-               } catch (Exception e) {
-                  throw new PersistenceException(e);
+         for (Map<K, FileEntry> segmentEntries : entries) {
+            if (segmentEntries == null)
+               continue;
+
+            List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
+            synchronized (segmentEntries) {
+               for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
+                  Map.Entry<K, FileEntry> next = it.next();
+                  FileEntry fe = next.getValue();
+                  if (fe.isExpired(now)) {
+                     it.remove();
+                     entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
```

Comment:
```suggestion
            List<Map.Entry<K, FileEntry>> entriesToPurge = new ArrayList<>();
            synchronized (segmentEntries) {
               for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
                  Map.Entry<K, FileEntry> next = it.next();
                  FileEntry fe = next.getValue();
                  if (fe.isExpired(now)) {
                     it.remove();
                     entriesToPurge.add(next);
```
This way we don't need to allocate the pairs.

Reply:
HashMap's entry iterator gives us the Node that's stored in the table, and I don't think it would be safe to read the value from it without synchronization.

Reply:
We hold the read lock, so it shouldn't be possible for the map to change though right?

Reply:
The more I think about it, can we not just use the stamped lock to protect the segmentEntries entry as well? This would reduce some code complexity as we don't have to worry about multiple locks when reading or writing it. Or is there a case where it wasn't protecting it?

Reply:
The problem is that we only have 1 `resizeLock` (which was initially used only to synchronize `allocate()` adding to `filePos` w/ `clear()` setting `filePos=0`, as the comment on the field still says), and if we hold it exclusively for every write then we lose the ability of writing to different segments in parallel. Granted, the parallel writes are not the main reason why people were using a segmented SFS before, but it seemed a shame to lose that.

So I only considered going the other way around: using `ConcurrentHashMap` for segment entries so that we can iterate over the entries without synchronization and hopefully avoid some of the large collection allocations we have here and elsewhere. But there I ran into the limitation that we use `LinkedHashMap` to implement `max-entries`. I figured I have a choice between 
1. Use Caffeine for the segment entries, which seemed excessive (but I admit I didn't look too much into it)
2. Keep using a synchronized `LinkedHashMap` when `max-entries` is set, but add a completely separate CHM-based path when `max-entries` is not set
3. (my choice) deprecate `max-entries` and keep using a synchronized `LinkedHashMap`/`HashMap` for now, move to CHM later.